File: problems/0001-0100/0032_group-tree-by-depth/solution_recursive_dfs.ts
function groupTreeByDepth(root: TreeNode | null): number[][] {
    // One list per depth, appended to the first time the walk reaches that
    // depth; afterwards it already exists for every later arrival.
    const grouped: number[][] = [];
    if (root === null || root === undefined) {
        return grouped;
    }
    // Pre-order: record the value before descending, so arrivals at each
    // depth happen left to right.
    const visit = (node: TreeNode, depth: number): void => {
        if (grouped.length === depth) {
            grouped.push([]);
        }
        grouped[depth]!.push(node.val);
        if (node.left !== null && node.left !== undefined) {
            visit(node.left, depth + 1);
        }
        if (node.right !== null && node.right !== undefined) {
            visit(node.right, depth + 1);
        }
    };
    visit(root, 0);
    return grouped;
}

File: problems-adapt/0001-0100/0032_group-tree-by-depth/solution_recursive_dfs.go
func groupTreeByDepth(root *TreeNode) [][]int {
	// One list per depth, appended to the first time the walk reaches that
	// depth; afterwards it already exists for every later arrival.
	grouped := [][]int{}
	var visit func(node *TreeNode, depth int)
	visit = func(node *TreeNode, depth int) {
		// Pre-order: record the value before descending, so arrivals at
		// each depth happen left to right.
		if len(grouped) == depth {
			grouped = append(grouped, []int{})
		}
		grouped[depth] = append(grouped[depth], node.Val)
		if node.Left != nil {
			visit(node.Left, depth+1)
		}
		if node.Right != nil {
			visit(node.Right, depth+1)
		}
	}
	if root != nil {
		visit(root, 0)
	}
	return grouped
}
